### PR TITLE
Read disc data asynchronously in a new thread

### PR DESCRIPTION
**Pull request by JosJuice**
DVDInterface inserts a delay when executing commands so that loading times will be similar to ones on consoles. During the delay, the CPU thread continues emulating the CPU. DVDInterface schedules a CoreTiming event so that it can execute the command after the delay. Executing the command involves reading data from the hard drive, a somewhat slow process... And it's happening on the CPU thread, so once the CoreTiming event starts running, CPU emulation needs to wait until DVDInterface is done.

This PR creates a new thread that reads data from the hard drive (and decompresses it, if a compressed disc image format is used) during the delay while the CPU thread is emulating the CPU. Copying the data to the emulated RAM is not done in this new thread; it's done in a CoreTiming event like before, so this PR shouldn't change the timing from the emulated system's point of view, and everything should still be deterministic. It could be useful to have an option to copy to RAM immediately after the data has been read, but I'm going to save that for a future PR.

@JMC47 Could you test the performance of this in a place where a lot of data is being read from the disc? FMVs, maybe? Make sure to have SUDTR off, because this isn't going to improve the performance of SUDTR. The difference in performance should be bigger if the ISO is on a slow storage device.

### CODE REVIEW

**Comment by magumagu** [comment]
It would be nice to have some sort of logging so we can compare how long a read takes in emulated time vs. real time it took.

**Comment by JosJuice** [comment]
With something like DEBUG_LOG? Sure, I can do that.

**Comment by JosJuice** [comment]
I've added logging (enable DVDInterface and Debug to see it), but the precision is apparently limited to 1 ms. Is std::chrono::steady_clock always this bad, or is it just a Windows problem? Are there any better options?

**Comment by NRo1e** [comment]
std::chrono::duration_cast<std::chrono::**micro**seconds> ---> std::chrono::duration_cast<std::chrono::**nano**seconds> perhaps? [ref](https://solarianprogrammer.com/2012/10/14/cpp-11-timing-code-performance/)

File: Source/Core/Core/HW/DVDThread.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+static void DVDThread();
+static bool ReadFromDisc();
+
+static void FinishRead(u64 userdata, int cyclesLate);
+static int g_finish_read;
+
+static std::thread g_DVD_thread;
+static std::mutex g_DVD_mutex;
+static std::condition_variable g_DVD_condition_working;
+static bool g_DVD_working = false;
```
Comment (lioncash) [inline Source/Core/Core/HW/DVDThread.cpp]: static variables don't need `g_`, but `s_`

**Comment by JosJuice** [comment]
The result when changing microseconds to nanoseconds is pretty much the same, except 1000 times larger.

**Comment by magumagu** [comment]
Common::Timer::GetTimeUs() has higher resolution.

**Comment by JosJuice** [comment]
That works well. Thanks!

**Comment by JMC47** [comment]
I did test this a bit.  I'm not sure if it was random, but when loading Metroid Prime (saved at the ship) I saw Samus already standing outside of the ship when it first loaded, then the cutscene played with her getting out of the ship.  Maybe that can happen in master too, or maybe it can even happen on console.

Otherwise, performance seemed fine.

**Comment by JosJuice** [comment]
I've also been testing Metroid Prime a bit, and I've found an obvious issue that seems to be consistently reproducible. First, start the game (GM8E01) as usual, then press Start+B+X to restart the game. Dolphin will display the following two panic alerts:

> Unknown Pointer 0x01800000 PC 0x813008ac LR 0x813006ac
> Invalid range in CopyToEmu. 14a0 bytes to 0x017feb60

After dismissing them, the game appears to be working normally at first, but there are several problems. When selecting a save file, the background music disappears. Once in-game, all audio is gone, and the HUD except for the health meter is also gone. The game also spawns you in the wrong room. My save was at the first save station in Phazon Mines, but I didn't start there. I started at the elevator going into Phazon Mines, with the elevator cutscene playing instead of the save station cutscene. When heading into the next room, the game acted as if it was my first time there, with the new area cutscene playing and the pirates present. My items and map were intact, though.

The CopyToEmu that triggers the panic alert is indeed the one in DVDThread. That CopyToEmu is pretty much the only thing that this PR changes in the DI-CPU interface. Earlier, DVDInterface used GetPointer and then asked DiscIO to write to that pointer. DVDThread now has its own buffer that it asks DiscIO to write to, and then it uses CopyToEmu at the end. I assume that the GetPointer method skips some error checking and thus somehow works, but I really don't know much more than that. It's technically possible to change DVDThread's code back to GetPointer plus a memcpy, but from what I understand, that's bad to do... Does anyone know what to do with this?

**Comment by magumagu** [comment]
That's a bug in CopyToEmu: the range ending at 0x01800000 is valid (even though 0x01800000 itself can't be dereferenced).  I guess in GetPointer(), `address < REALRAM_SIZE` should be `address <= REALRAM_SIZE`.

**Comment by JosJuice** [comment]
That seems to do the trick. I've added a commit with that fix.

**Comment by comex** [comment]
I was going to ask about using fadvise and equivalent instead, but I guess that wouldn't help with compression.

**Comment by JosJuice** [comment]
@comex I've changed std::condition_variable to Common::Event, but there is now sometimes a deadlock in which both the CPU thread (in FinishRead) and the DVD thread are running `s_DVD_thread_state_changed.Wait()` at the same time. I don't know exactly how that can happen, but could it be because there no longer is a mutex for the while loop? I got rid of it since it otherwise would be impossible to do any work if a thread is waiting.

**Comment by shuffle2** [comment]
Please make sure the gamecube service disc still works after these changes.

**Comment by shuffle2** [comment]
Also I agree with @comex that there probably is a better way to do this. Either use an OS primitive to inform how read-aheads should be done (I think it should actually work fine for compression, since compressed blocks maintain same order as physical disc), or use a cache in dolphin (probably easier to have a good cross-platform solution).

**Comment by JosJuice** [comment]
The service disc works in general, but the Optical Disc Drive Test still fails like it has done for at least 4000 versions. https://code.google.com/p/dolphin-emu/issues/detail?id=8358

Doing read-aheads like that would work fine for reading the compressed data from the disk, but it wouldn't help with decompressing it. That's acceptable performance-wise for a typical use case like decrypting Wii data and decompressing a gcz, but I've been thinking of adding support for WIA in the future, and it could be a performance problem for formats with heavy compression like that.

How does a cache solve the problem? It would speed up loading data that has been loaded recently, but not any of the other data. (Unless the entire ISO is cached, I suppose, but that would be a huge waste of RAM...)

**Comment by JosJuice** [comment]
I've changed how this works a little, and the deadlocks appear to be gone. Instead of having a flag that tells if the thread is working and an event that tells when the flag is changed, there are two events, one that tells the DVD thread to start and one that tells the CPU thread to stop. This feels cleaner in general to me, but WaitUntilIdle is rather awkward. Am I going the wrong way about that? If not, maybe I could add some sort of WaitWithoutClearing method to Event?

**Comment by shuffle2** [comment]
btw the timing stuff should have been fixed by VS2015

File: Source/Core/Core/HW/DVDInterface.cpp [hunk 2]
```diff
@@ -329,32 +311,11 @@ static void FinishExecuteCommand(u64 userdata, int cyclesLate)
...
-	current_read_command.is_valid = false;
-
-	// The final step is to notify the emulated software that the command has been executed
-	CoreTiming::ScheduleEvent_Immediate(current_read_command.callback_event_type,
-	                                    current_read_command.interrupt_type);
-}
-
 static u32 ProcessDTKSamples(short *tempPCM, u32 num_samples)
 {
+	// TODO: Read audio data using the DVD thread instead of blocking on it?
```
Comment (degasus) [inline Source/Core/Core/HW/DVDInterface.cpp]: Yes, just push a streaming command to the dvd thread for it. Maybe also with the coretiming value as we'll need it for good audio dumping.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	if (s_DVD_success)
+		Memory::CopyToEmu(s_output_address, s_DVD_buffer.data(), s_length);
+
+	// Notify the emulated software that the command has been executed
+	CoreTiming::ScheduleEvent_Immediate(s_callback_event_type, DVDInterface::INT_TCINT);
+}
+
+void DVDThread()
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Functions which are designed to not be called from extern should be marked as static.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	// Set the event again so that we still know that the DVD thread isn't working
+	s_DVD_thread_not_working.Set();
+}
+
+void StartRead(u64 DVD_offset, u32 output_address, u32 length, bool decrypt,
+               int callback_event_type, int ticks_until_completion)
+{
+	_assert_(Core::IsCPUThread());
+
+	s_DVD_thread_not_working.Wait();
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Are games able to queue more than one commands at once? If so, we should use a queue here instead.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			break;
+		else
+			s_DVD_success = ReadFromDisc();
+	}
+
+	if (s_DVD_success)
+		Memory::CopyToEmu(s_output_address, s_DVD_buffer.data(), s_length);
+
+	// Notify the emulated software that the command has been executed
+	CoreTiming::ScheduleEvent_Immediate(s_callback_event_type, DVDInterface::INT_TCINT);
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Maybe a bit out of scope of this PR, but isn't this just a generic callback? So instead of the "int callback_event_type" everywhere, you can just use a "std::function callback".

File: Source/Core/Core/HW/DVDInterface.cpp (hunk 2 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDInterface.cpp]: It's not as simple as just pushing a command to the DVD thread, because it can only handle one command at once. I'd prefer to leave it to another PR so that this one doesn't get more complicated than it already is.

What do you mean by the last sentence?

File: Source/Core/Core/HW/DVDThread.cpp (hunk 4 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: There is a queue in IOS, which we appropriately handle in WII_IPC_HLE. DI doesn't have a queue (except for DTK, I guess).

File: Source/Core/Core/HW/DVDThread.cpp (hunk 5 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: Putting a regular callback in savestates doesn't work well.

File: Source/Core/Core/HW/Memmap.cpp [hunk 6]
```diff
@@ -302,12 +302,12 @@ u8* GetPointer(u32 address)
 	// TODO: Should we be masking off more bits here?  Can all devices access
 	// EXRAM?
 	address &= 0x3FFFFFFF;
-	if (address < REALRAM_SIZE)
+	if (address <= REALRAM_SIZE)
```
Comment (degasus) [inline Source/Core/Core/HW/Memmap.cpp]: This is a big hack, and in my opinion, broken. if address is equal to REALRAM_SIZE, the pointer will be invalid.

Comment (phire) [inline Source/Core/Core/HW/Memmap.cpp]: The correct fix would be to subtract one from the call in ValidCopyRange

File: Source/Core/Core/HW/Memmap.cpp [hunk 7]
```diff
@@ -247,8 +247,8 @@ bool AreMemoryBreakpointsActivated()
 static inline bool ValidCopyRange(u32 address, size_t size)
 {
 	return (GetPointer(address) != nullptr &&
-	        GetPointer(address + u32(size)) != nullptr &&
-	        size < EXRAM_SIZE); // Make sure we don't have a range spanning seperate 2 banks
+	        GetPointer(address + u32(size) - 1) != nullptr &&
+	        size <= EXRAM_SIZE); // Make sure we don't have a range spanning 2 separate banks
```
Comment (degasus) [inline Source/Core/Core/HW/Memmap.cpp]: "size < EXRAM_SIZE" sounds fine.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+static void FinishRead(u64 userdata, int cyclesLate)
+{
+	WaitUntilIdle();
+
+	DEBUG_LOG(DVDINTERFACE, "Disc has been read. Real time: %llu us. Real time including delay: %llu us. Emulated time including delay: %llu us.",
+	          s_realtime_done_us - s_realtime_started_us,
+	          Common::Timer::GetTimeUs() - s_realtime_started_us,
+	          (CoreTiming::GetTicks() - s_time_read_started) / (SystemTimers::GetTicksPerSecond() / 1000 / 1000));
+
+	while (!s_DVD_success)
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Do you have to implement this loop on the CPU thread? Else it would likely be cleaner within the ReadFromDisc function itself on the dvd thread.

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: I don't want emulation to continue running when the user is being asked the question. Doing this on the DVD thread isn't especially bad right now, because the CPU thread will stop and wait for the DVD thread after a short time no matter what, but it will matter in the future with my planned option to only call FinishRead when the DVD thread is done.

**Comment by degasus** [comment]
I've checked out the "native" OS ways to do this on linux: There is the AIO framework which sounds like exactly what we wants here, but with lots of pitfalls. eg the read must match the block size, ..., and the worst, those IOs bypass the cache. Still possible, but I'm not sure if this fits well.
The other way is the official posix way. But glibc will just spawn a new thread per IO request. So using our own worker thread should be much faster.

I have no clue about windows, but this sounds like just a generic implementation with a worker thread might be very fine...

File: Source/Core/Core/HW/DVDThread.cpp [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	s_DVD_thread_start_working.Set();
+
+	CoreTiming::ScheduleEvent(ticks_until_completion, s_finish_read);
+}
+
+static void FinishRead(u64 userdata, int cyclesLate)
+{
+	WaitUntilIdle();
+
+	DEBUG_LOG(DVDINTERFACE, "Disc has been read. Real time: %llu us. Real time including delay: %llu us. Emulated time including delay: %llu us.",
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: "%" PRIu64 instead of "%llu"

File: Source/Core/Core/HW/DVDThread.cpp [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+static void DVDThread();
+static bool ReadFromDisc();
+
+static void FinishRead(u64 userdata, int cyclesLate);
+static int s_finish_read;
+
+static std::thread s_DVD_thread;
+static Common::Event s_DVD_thread_start_working;
+static Common::Event s_DVD_thread_not_working;
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: I think some kind of semaphore will match better than two events. eg you can map the number of read commands queued for the worker. So WaitUntilIdle will block until this number hits zero. StartRead will increment it. And the worker thread will wait for a number bigger than 0 and decrement it when the work is done.
In the end, a semaphore needs as many ressources as one event (condition variable + mutex), but it's likely faster to call WaitUntilIdle redundantly.

But bad luck, we don't have a semaphore within Common, so feel free to ignore me.

**Comment by degasus** [comment]
I've tested this code quite a bit today, and it seems to work just great. I didn't expect dolphin to be blocked for 10ms often just for disk access, but we are. At least here on my hard disk.

To be honest, I guess it may also be worth to delay it even more, eg as long as the real harddisk need.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+		if (PanicYesNoT("The disc could not be read (at 0x%" PRIx64 " - 0x%" PRIx64 ").\n"
+		                "Do you want to continue without retrying?",
+		                s_DVD_offset, s_DVD_offset + s_length))
+			break;
+		else
+			s_DVD_success = ReadFromDisc();
+	}
+
+	if (s_DVD_success)
+		Memory::CopyToEmu(s_output_address, s_DVD_buffer.data(), s_length);
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: This code runs on the CPU thread, and it's a big memcpy (here, 10% of the cpu time of ReadFromDisc). I see there is a need to have a deterministic way for netplay, but else it may be faster to just memcpy from the dvd thread into gc memory.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+	s_DVD_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_DVD_buffer);
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Maybe just store the command to read the dvd? So you'll just re-read the dvd after loading. It's a bit in contrast to WaitUntilIdle() before, but it may work quite fine.

Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Else you may also just call s_DVD_buffer.resize(0); after you've used it. This won't free the memory, but so it won't be included within the savestate often.

File: Source/Core/Core/HW/DVDThread.cpp (hunk 11 above)
Comment (phire) [inline Source/Core/Core/HW/DVDThread.cpp]: Faster, yes. 

Bad idea, yes.

Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Why? It matches better to the real hardware. But of course, the timing is worse.

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: Doing the memcpy too early could lead to the problem described in PR #1854. Because the DVD thread doesn't run at the same speed as the real hardware, we can't rely on its timing to be accurate.

Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: Will your planned "fast disc access" PR also cover this? If we're going to set those interrupts on the dvd thread, then there won't be any need to not do the memcpy on the dvd thread, too.

File: Source/Core/Core/HW/DVDThread.cpp [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+	// s_realtime_started_us and s_realtime_done_us aren't savestated
+	// because they rely on the current system's time.
+	// This means that loading a savestate might cause
+	// incorrect times to be logged once.
+}
+
+void WaitUntilIdle()
+{
+	_assert_(Core::IsCPUThread());
```
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: This assert is hit with savestates. They aren't called from the CPU thread

#8  0x00000000005f1ead in DVDThread::WaitUntilIdle () at ../Source/Core/Core/HW/DVDThread.cpp:105
#9  0x00000000005f2389 in DVDThread::DoState (p=...) at ../Source/Core/Core/HW/DVDThread.cpp:83
#10 0x0000000000601783 in HW::DoState (p=...) at ../Source/Core/Core/HW/HW.cpp:97
#11 0x00000000005c5444 in State::DoState (p=...) at ../Source/Core/Core/State.cpp:172
#12 0x00000000005c6bf4 in State::SaveToBuffer (buffer=std::vector of length 0, capacity 0) at ../Source/Core/Core/State.cpp:204
#13 0x00000000005c7bc7 in State::LoadAs (filename="/home/markus/.dolphin-emu/StateSaves/GE4E7D.s01") at ../Source/Core/Core/State.cpp:518
#14 0x00000000005c807a in State::LoadAs (filename="/home/markus/.dolphin-emu/StateSaves/GE4E7D.s01") at ../Source/Core/Core/State.cpp:641
#15 State::Load (slot=slot@entry=1) at ../Source/Core/Core/State.cpp:640
#16 0x000000000050bf1b in CFrame::ParseHotkeys (this=0x18f3cc0) at ../Source/Core/DolphinWX/Frame.cpp:1513
#17 0x00007ffff21d393e in wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const () from /usr/lib/libwx_baseu-3.0.so.0
#18 0x00007ffff236b2e8 in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#19 0x00007ffff236b6de in wxEvtHandler::SearchDynamicEventTable(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#20 0x00007ffff236b76e in wxEvtHandler::TryHereOnly(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#21 0x00007ffff236b833 in wxEvtHandler::ProcessEventLocally(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#22 0x00007ffff236b895 in wxEvtHandler::ProcessEvent(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#23 0x00007ffff236b5f7 in wxEvtHandler::SafelyProcessEvent(wxEvent&) () from /usr/lib/libwx_baseu-3.0.so.0
#24 0x00007ffff22cb071 in wxTimerImpl::SendEvent() () from /usr/lib/libwx_baseu-3.0.so.0
#25 0x00007ffff2f3e1ab in ?? () from /usr/lib/libwx_gtk2u_core-3.0.so.0
#26 0x00007ffff58526e3 in ?? () from /usr/lib/libglib-2.0.so.0
#27 0x00007ffff5851c7a in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#28 0x00007ffff5852020 in ?? () from /usr/lib/libglib-2.0.so.0
#29 0x00007ffff5852342 in g_main_loop_run () from /usr/lib/libglib-2.0.so.0
#30 0x00007ffff3676467 in gtk_main () from /usr/lib/libgtk-x11-2.0.so.0
#31 0x00007ffff2f1f485 in wxGUIEventLoop::DoRun() () from /usr/lib/libwx_gtk2u_core-3.0.so.0
#32 0x00007ffff2218c53 in wxEventLoopBase::Run() () from /usr/lib/libwx_baseu-3.0.so.0
#33 0x00007ffff21d7b46 in wxAppConsoleBase::MainLoop() () from /usr/lib/libwx_baseu-3.0.so.0
#34 0x00007ffff226e8f0 in wxEntry(int&, wchar_t**) () from /usr/lib/libwx_baseu-3.0.so.0
#35 0x000000000043e268 in main (argc=1, argv=<optimized out>) at ../Source/Core/DolphinWX/Main.cpp:80

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: We call PauseAndLock before DoState, so there is no thread safety issue, only the issue of the assert being triggered incorrectly. PR #3101's changes to PauseAndLock should fix this by declaring the thread that locks as the CPU thread. Should we wait for PR #3101, or should I include the PauseAndLock part of it in this PR?

File: Source/Core/Core/HW/DVDThread.cpp (hunk 11 above)
Comment (phire) [inline Source/Core/Core/HW/DVDThread.cpp]: We really need it to be deterministic over anything else. Determinism isn't just useful for netplay; Bugs are hard enough to find when they are reproducible, it gets a lot harder when timings can vary based on on your computer's setup or other factors.

The DVD could start writing data instantly if the data was already in the OS's cache, potentially before the CPU thread even continues execution. Or it could hit a spinning HDD and take 10-12ms.

Maybe if we had locking (hint, hint me) we could lock a region of memory and allow other threads to write into main memory while maintaining determinism.

File: Source/Core/Core/HW/DVDThread.cpp (hunk 13 above)
Comment (degasus) [inline Source/Core/Core/HW/DVDThread.cpp]: I'd just rebase

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: I've rebased on latest master which includes PR #3101.

File: Source/Core/Core/HW/DVDThread.cpp (hunk 12 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: I've added a `resize(0)` for now. The first suggestion saves more space in savestates, but I'll save it for some other time (maybe if I change DVDThread to use a queue in the future)...

File: Source/Core/Core/HW/DVDThread.cpp (hunk 11 above)
Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: @degasus If the callback that sets the interrupt is scheduled using the DVD thread, then doing the memcpy on the DVD thread would indeed make sense, but I don't want that callback to always be scheduled on the DVD thread when the option is enabled. Because of the problems mentioned above, I'd rather let users set something like "let disc reads finish up to 300% as fast" than just having "let disc reads finish up to infinitely fast". My plan for this is to have the CPU thread schedule a "earliest allowed completion time" event (and a "latest allowed completion time" event?) and to have the DVD thread schedule an "I'm done" event when it's done, with the events doing what's appropriate depending on the order they're called. I suppose I can add some sort of way for the DVD thread to check if the "earliest allowed completion time" event has been called yet so that it knows if it can do memcpy on its own, but either way, this is something that won't be in this PR.

**Comment by degasus** [comment]
LGTM

File: Source/Core/Core/HW/DVDThread.cpp [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+	          "Real time including delay: %" PRIu64 " us. Emulated time including delay: %" PRIu64 " us.",
+	          s_realtime_done_us - s_realtime_started_us,
+	          Common::Timer::GetTimeUs() - s_realtime_started_us,
+	          (CoreTiming::GetTicks() - s_time_read_started) / (SystemTimers::GetTicksPerSecond() / 1000 / 1000));
+
+	while (!s_dvd_success)
+	{
+		// Asking "Do you want to retry?" would be more user friendly,
+		// but if panic alerts are turned off, the answer is automatically Yes,
+		// and we don't want an infinite loop where ReadFromDisc always fails...
```
Comment (phire) [inline Source/Core/Core/HW/DVDThread.cpp]: A) I'm really not happy with you relying on this behaviour.
B) as of #3241, The no GUI build of dolphin always returns No, if panic handlers are enabled.

Comment (phire) [inline Source/Core/Core/HW/DVDThread.cpp]: Not even sure why we would need a retry feature. In Boot you don't even check the return code.

The only time I've ever seen a read failure was when I had problems with my locking branch and the memory permissions were wrong due to a race condition.

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: The case I was thinking of is someone storing games on external media and accidentally unplugging it. It doesn't happen often, but giving the user a chance to recover is better than desyncing and very likely making the game crash. Boot not caring about return values is an argument for fixing Boot, not for not having this.

Because adding a retry feature doesn't really have anything to do with making disc reads asynchronous and because there are issues with it, I will drop it from this PR. I might end up submitting it as a separate PR later in the future if I manage to make it cleaner (or soon if there's someone who's interested in this feature).

Comment (phire) [inline Source/Core/Core/HW/DVDThread.cpp]: I guess that's not a bad idea. But put it in another PR

Comment (JosJuice) [inline Source/Core/Core/HW/DVDThread.cpp]: It has been removed now.
